c7n/resources/s3.py

```diff
@@ -783,6 +785,165 @@ def process_bucket(self, b):
         return None
 
 
+@filters.register('has-policy-permission')
+class HasPermissionFilter(Filter):
+    """Filters buckets with certain permissions.
+
+    A bucket is filtered if any of the given actions and any of the given principals match
+    the bucket's policy.
+
+    This allows matching bucket policies that contain wildcards.  For example, if you
+    specify `s3:PutObject` in a filter and a bucket policy has a `s3:Put*` action, the
+    bucket will match.
+
+    Conversely, if you specify a wildcard in a filter (e.g `*`), this will match all bucket
+    policy permissions.
+
+    Special considerations:
+
+    * Principals should be specified as "prefix:value" where prefix is one of AWS,
+      Federated, Service, etc.
```

Comment:
nice docs, i'd probably say reference principal in iam documentation via url here.